Asset tracking: line-simplification for geo_line in TSDB

craigtaverner opened the pull request:
Part of the work described in https://github.com/elastic/elasticsearch/issues/85274

Within time-series data, two optimisations are done:

- [x] Optimizations based on TSDB data ordering (tsid and time-series order)
- [x] Optimizations based on line simplification from https://github.com/elastic/elasticsearch/pull/94859

Tasks to complete before merging:

- [x] Yaml tests for time-series
  - [x] Basic tests sorting by tsid ASC and DESC
  - [x] Basic test with terms agg on same field as tsid
  - [x] test for multiple terms and TSID
  - [x] Negative tests (disallowed sort field)
- [x] Unit tests for time-series
  - [x] Test ASC and DESC for single TSID
  - [x] test ASC and DESC for terms on TSID
  - [x] Test for multiple TSID
  - [x] test for multiple terms and TSID

Final decision on when to activate optimized geo_line:

  - [x] geo_line inside time_series (required)
  - [ ] ~~geo_line on position metric~~, decided to not require this, and instead:
    - [x] Automatically default sort field to `@timestamp`
    - [x] Throw exception on any other sort field setting

So now the switch between old and new geo_line behaviour is exclusively based on its nesting inside a time_series aggregation.

Work to be done in additional PRs:

- [ ] ~~Use registered `ValueSource->Aggregator` mappings to map position metrics to `GeoLineAggregator.TimeSeries`~~(cancelled based on decision above)
- [x] Optimize `MergedGeoLines` for time-series data (ie. append and re-simplify) - https://github.com/elastic/elasticsearch/pull/96953
- [ ] Allow user to turn on line simplification without time-series (geo_line flag): https://github.com/elastic/elasticsearch/issues/87903
- [ ] Allow user to specify geo_bounds within geo_line (simplify kibana queries)

craigtaverner committed:
WIP Started geo_line for TSDB work

Starting with YAML tests (which currently pass) and AggregatorTests
(currently failing, likely due to mistake in the tests)

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Update docs/changelog/94954.yaml

craigtaverner committed:
WIP Refactoring to prepare for TSDB geo_line

* Created TimeSeries version of GeoLineAggregator, and wired it in so that time-series aggregations use it, but current behavior is still identical to non-time-series.
* Added both yaml and unit tests for testing that geo_line works with correct results in both time-series and non-time-series cases.
* Added additional tests to verify the grouping behaviour of time-series vs. terms aggs, and the combination of the two.

craigtaverner committed:
WIP Refactoring to prepare for TSDB geo_line

* Started refactoring to re-use simplifier for all buckets

craigtaverner committed:
Fixed bug with leaf collector not changing per segment

craigtaverner committed:
Fixed bug with leaf collector not detecting bucket changes

The bucket id can change within a segment, so we need to detect this and save the geo_line.

craigtaverner committed:
Renamed class since it no longer extends BucketedSort

The original geo_line relied on the BucketedSort for all intelligence.
The time-series geo_line uses none of that, and does its own memory management.

elasticsearchmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

craigtaverner committed:
Fixed bug with geo_point leaking between geo_line buckets

And enhanced unit tests to cover multiple groups

iverase commented on test/framework/src/main/java/org/elasticsearch/search/aggregations/AggregatorTestCase.java:
```diff
@@ -467,7 +467,7 @@ protected <A extends InternalAggregation, C extends Aggregator> A searchAndReduc
         IndexSettings indexSettings = createIndexSettings();
         // First run it to find circuit breaker leaks on the aggregator
-        runWithCrankyCircuitBreaker(indexSettings, searcher, aggTestConfig);
+        // runWithCrankyCircuitBreaker(indexSettings, searcher, aggTestConfig);
```
I guess this is a left over?

craigtaverner replied:
Good catch. I repeatedly commented that out and uncommented it in order to make debugging easier, but clearly forgot to uncomment it the last time!

iverase commented on x-pack/plugin/spatial/build.gradle:
```diff
@@ -12,6 +12,7 @@ esplugin {
 
 dependencies {
   compileOnly project(path: ':modules:legacy-geo')
+  compileOnly project(path: ':modules:aggregations')
```
I think this is only used in test, testImplementation would be better?

craigtaverner replied:
Done

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/GeoLineAggregator.java:
```diff
@@ -6,6 +6,8 @@
  */
 package org.elasticsearch.xpack.spatial.search.aggregations;
 
+import com.unboundid.util.NotNull;
```
I guess if we could avoid using this by just asserting that the input is not null as we are doing in other places.

craigtaverner replied:
I just checked and we only call these constructors from code that always creates the value source, so no check is needed.

My addition of the null check was as a result of refactoring code that had lots of `if(valueSource==null){}` checks, but that code was supporting both Empty and non-Empty cases, which is no longer the case. So now I just removed the checks.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/GeoLineAggregator.java:
```diff
@@ -50,65 +48,160 @@ final class GeoLineAggregator extends MetricsAggregator {
-    @Override
-    public void doClose() {
-        Releasables.close(sort, extra, counts);
+    static class TimeSeries extends GeoLineAggregator {
```
Maybe a better name would be `TimeSeriesGeoLineAggregator`?

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/TimeSeriesGeoLineBuckets.java:
```diff
@@ -0,0 +1,328 @@
+        boolean complete = simplifier.length() < bucketSize;
+        if (line.isEmpty()) {
+            // TODO: Perhaps we should return an empty geo_line here?
+            throw new IllegalStateException("Invalid bucket " + bucket + " for geo_line");
```
yes, I think we should return an empty aggregation instead of throwing an error

craigtaverner replied:
Done

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/GeoLineAggregatorFactory.java:
```diff
@@ -58,7 +58,11 @@ protected Aggregator doCreateInternal(
     ) throws IOException {
         GeoLineMultiValuesSource valuesSources = new GeoLineMultiValuesSource(configs);
-        return new GeoLineAggregator(name, valuesSources, context, parent, metaData, includeSort, sortOrder, size);
+        if (context.isInSortOrderExecutionRequired()) {
```
Are we checking somewhere else that the provided sort is @timestamp? if it is not then we should not use this path?

craigtaverner replied:
You are right, I do not yet do the sort-field check. I had a comment in the description about missing negative tests, but for sure this is also missing code, so definitely needs to be added.

craigtaverner replied:
OK. Now this is done. I check that the sort field is `@timestamp` the hard-coded field for time-series data stream, and only if that is true allow the geo_line optimization. I also enhanced the tests to run with the field as either `@timestamp` or `time_field` and verify different behaviour for the two cases.

For `position` metrics, we can disallow the sort-field entirely. That can be a followup PR.

iverase reviewed (changes requested):
I left a few comments, it is looking good. My only concerns at the moment is about sort validation on the aggregation when running in time series mode. In particular we discussed the possibility on not defining it for time series so it is implicit?

craigtaverner commented:
Thanks for the review. I'll certainly add the missing sort-field checks and associated tests, and make other fixes. Probably only next week though.

craigtaverner committed:
Code review updates

craigtaverner committed:
Verify that the sort field is specifically the TS timestamp

Only activate the time-series optimizations if the aggregation is both:
* Within a time-series aggregation (ie. tsid and @timestamp ordered)
* The geo_line sort field is @timestamp

craigtaverner committed:
Allow geo_point time-series to skip sort config

Also disables the new geo_line for time-series even if the correct
sort and point fields are used if the point field is not explicitly
configured to be a position metric.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/search/aggregations/GeoLineAggregationBuilder.java:
```diff
@@ -151,12 +154,43 @@ protected MultiValuesSourceAggregatorFactory innerBuild(
+                    defaultValueSourceType()
+                );
+                configs.put(SORT_FIELD.getPreferredName(), sourceConfig);
+            } else if (false && sourceConfig.fieldContext().field().equals(TIMESTAMP_FIELD.getName()) == false) {
```
This is always false?

iverase replied:
Let's do that as a follow up, this PR is already pretty big.

craigtaverner committed:
Support geo_centroid and geo_bounds on position metric

craigtaverner committed:
Update yaml tests for multi-terms tests

craigtaverner committed:
Changed to disallow alternative sort-fields in ts-geo_line

Since the primary criteria for switching to the new algorithm is that
geo_line is within a time-series aggregation, we now disallow any other sort field.

We test the negative case in the yaml tests, but changed the unit tests to
use TermsAggregation to minim the time-series aggregation to get comparable
results.

iverase reviewed (approved):
LGTM

craigtaverner committed:
For non-time-series check missing sort field early

The old code only threw error if there was data because the check was done
inside the leaf collector just before actually reading the sort field.
And there were no tests for missing sort field.

This commit adds the tests, and checks early so even if data is missing.

craigtaverner committed:
Reviewed TODOs

* Test that behaviour is identical with or without POSITION metric
* Removed fallback code in builder (was switching to old geo_line without POSITION metric)
* Removed two TODO's that are no longer valid concerns

elasticsearchmachine commented:
@craigtaverner according to this PR's labels, I need to update the changelog YAML, but I can't because the PR is closed. Please either update the changelog yourself on the appropriate branch, or adjust the labels. Specifically:

   * The PR is labelled `release highlight` but the changelog has no `highlight` section
